### PR TITLE
team-level std algos: part 3

### PR DESCRIPTION
**Pull request by fnrizzi**
Co-authored-by: Cezary Skrzyński
Co-authored-by: Jakub Strzebonski

## Content 

Team-level implementation and tests for:

- `Kokkos_Find.hpp`
- `Kokkos_FindIf.hpp` 
- `Kokkos_FindIfNot.hpp` 
- `Kokkos_AllOf.hpp`
- `Kokkos_AnyOf.hpp`
- `Kokkos_NoneOf.hpp`
- `Kokkos_SearchN.hpp`

## Technical comments

-  this PR is a single one because of the following dependencies: 
   - `Kokkos_{AllOf, AnyOf, NonOf}.hpp`:  all need `find_if`
   - `Kokkos_SearchN.hpp`: needs `all_of`

- the change `class` -> `typename` and the enable if defaulted in the template parameters comes from [this request](https://github.com/kokkos/kokkos/pull/6200#issuecomment-1626406816)

- the code duplication is intentional and agreed upon after the long discussion in [here](https://github.com/kokkos/kokkos/pull/6200)

- tests are run for both contiguous views and strided views

## Other comments

- all team-level PRs are tracked in: https://github.com/kokkos/kokkos/issues/6352

- this comes from breaking apart https://github.com/kokkos/kokkos/pull/5150

## Possible conflicts

Implementation-wise, this has not conflict with other team-level PRs so it could be merged independently. 
However, a trivial conflict in the CMakeLists inside the unit test occurs if another of the team-level PRs is merged before this because each team-level PR adds test using a specific "label". Resolving this is trivial. 

## Associated wiki PRs

- `Kokkos_Find.hpp`, `Kokkos_FindIf.hpp`, `Kokkos_FindIfNot.hpp`: see [wiki PR](https://github.com/kokkos/kokkos-core-wiki/pull/366)
- `Kokkos_AllOf.hpp`, `Kokkos_AnyOf.hpp`, `Kokkos_NoneOf.hpp`: see [wiki PR](https://github.com/kokkos/kokkos-core-wiki/pull/364)
- `Kokkos_SearchN.hpp`: see [wiki PR](https://github.com/kokkos/kokkos-core-wiki/pull/371)

IMPORTANT: wiki PRs are not complete yet but 99% there 

The issue tracking all *wiki PRs* is [here](https://github.com/kokkos/kokkos-core-wiki/issues/379)

### CODE REVIEW

**Comment by fnrizzi** [comment]
CI error:
```
[ 91%] Building CXX object algorithms/unit_tests/CMakeFiles/Kokkos_UnitTest_StdSet_Team_C.dir/TestStdAlgorithmsTeamAllOf.cpp.o
NVC++-S-0000-Internal compiler error. extractor: bad opc       0 
(/var/jenkins/workspace/Kokkos/algorithms/unit_tests/TestStdAlgorithmsTeamFindIfNot.cpp: 216)
NVC++-S-0000-Internal compiler error. extractor: bad opc       0  (/var/jenkins/workspace/Kokkos/algorithms/unit_tests/TestStdAlgorithmsTeamFindIfNot.cpp: 216)
NVC++-S-0000-Internal compiler error. extractor: bad opc       0  (/var/jenkins/workspace/Kokkos/algorithms/unit_tests/TestStdAlgorithmsTeamFindIfNot.cpp: 216)
NVC++-S-0000-Internal compiler error. extractor: bad opc       0  (/var/jenkins/workspace/Kokkos/algorithms/unit_tests/TestStdAlgorithmsTeamFindIfNot.cpp: 216)
NVC++-S-0000-Internal compiler error. extractor: bad opc       0  (/var/jenkins/workspace/Kokkos/algorithms/unit_tests/TestStdAlgorithmsTeamFindIfNot.cpp: 216)
```

**Comment by fnrizzi** [comment]
retest this please

**Comment by fnrizzi** [comment]
CI issues:
- openmptarget (unrelated):
```
The following tests FAILED:
	  1 - Kokkos_CoreUnitTest_Serial1 (Timeout)
	  4 - Kokkos_CoreUnitTest_OpenMPTarget (Timeout)
	 18 - Kokkos_CoreUnitTest_StackTraceTest (Timeout)
```
	 
- nvhpc: not sure what is going on, timesout but i can build and run just fine on kokkos-dev-2

File: header [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+//               Solutions of Sandia, LLC (NTESS).
+//
+// Under the terms of Contract DE-NA0003525 with NTESS,
+// the U.S. Government retains certain rights in this software.
+//
+// Part of Kokkos, under the Apache License v2.0 with LLVM Exceptions.
+// See https://kokkos.org/LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//@HEADER
```
Comment (Rombur) [inline header:15]: This should be removed. Same with `header_diff`

File: algorithms/unit_tests/CMakeLists.txt [hunk 2]
```diff
@@ -173,6 +190,23 @@ foreach(Tag Threads;Serial;OpenMP;Cuda;HPX;HIP;SYCL;OpenMPTarget)
   endif()
 endforeach()
 
+
+# need to remove tests for OpenMPTarget because in these cases
+# the impl needs to use either Kokkos or tailored reducers
+# which results in runtime memory errors.
```
Comment (Rombur) [inline algorithms/unit_tests/CMakeLists.txt:196]: Add `FIXME_OPENMPTARGET`

**Comment by fnrizzi** [comment]
only hvhpc fails with compiler error opc

**Comment by Rombur** [comment]
There is a conflict with develop

**Comment by fnrizzi** [comment]
> There is a conflict with develop

yes i know, fixing it now - this will happen everytime see: 

```
Implementation-wise, this has not conflict with other team-level PRs so it could be merged independently.
However, a trivial conflict in the CMakeLists inside the unit test occurs if another of the team-level PRs is merged before this because each team-level PR adds test using a specific "label". Resolving this is trivial.
```

**Comment by fnrizzi** [comment]
> There is a conflict with develop

thanks for point it out!

**Review by Rombur — APPROVED** [review]

**Review by crtrott — APPROVED** [review]

**Pull request merged**